armeria-backend/src/main/scala/sttp/client3/armeria/ArmeriaBackend.scala

```diff
@@ -0,0 +1,201 @@
+import sttp.capabilities
+import sttp.client3.SttpClientException.ReadException
+import sttp.client3.armeria.ArmeriaBackend.PseudoHeaderPrefix
+import sttp.client3.internal.{BodyFromResponseAs, FileHelpers, SttpFile}
+import sttp.client3.ws.{GotAWebSocketException, NotAWebSocketException}
+import sttp.client3.{
+  ByteArrayBody,
+  ByteBufferBody,
+  FileBody,
+  FollowRedirectsBackend,
+  InputStreamBody,
+  MultipartBody,
+  NoBody,
+  Request,
+  Response,
+  StreamBody,
+  StringBody,
+  SttpBackend,
+  SttpClientException,
+  WebSocketResponseAs
+}
+import sttp.model._
+import sttp.monad.{FutureMonad, MonadError}
+
+import java.io._
+import java.nio.charset.StandardCharsets.UTF_8
+import java.nio.file.Files
+import java.util
+import java.util.Map.Entry
+import java.util.zip.{GZIPInputStream, InflaterInputStream}
+import scala.annotation.tailrec
+import scala.concurrent.{ExecutionContext, Future}
+import scala.jdk.FutureConverters._
+import scala.util.Try
+
+class ArmeriaBackend(client: Option[WebClient] = None)(implicit ec: ExecutionContext = ExecutionContext.global)
+    extends SttpBackend[Future, Any] {
+  type PE = Any with capabilities.Effect[Future]
+
+  override def send[T, R >: PE](request: Request[T, R]): Future[Response[T]] =
+    adjustExceptions(request)(execute(request))
+
+  private def execute[T, R >: PE](request: Request[T, R]): Future[Response[T]] =
+    getClient(request)
+      .execute(requestToArmeria(request))
+      .aggregate()
+      .asScala
+      .flatMap(responseFromArmeria(request, _))
+
+  private def getClient(request: Request[_, _]): WebClient =
+    client.getOrElse(
+      WebClient
+        .builder()
+        .responseTimeoutMillis(request.options.readTimeout.toMillis)
+        .build()
+    )
+
+  private def adjustExceptions[T](request: Request[_, _])(execute: => Future[T]): Future[T] =
+    SttpClientException.adjustExceptions(responseMonad)(execute) {
+      case ex @ (_: ClosedSessionException | _: ResponseTimeoutException) =>
+        Some(new ReadException(request, ex))
+      case ex =>
+        SttpClientException.defaultExceptionToSttpClientException(request, ex)
+    }
+
+  private def requestToArmeria(request: Request[_, Nothing]): HttpRequest = {
+    val method = methodToArmeria(request.method)
+    val path = request.uri.toString()
+    val headers = headersToArmeria(request.headers, method, path)
+
+    (request.body match {
+      case NoBody                 => HttpRequest.of(method, path)
+      case StringBody(s, _, _)    => HttpRequest.of(method, path, PLAIN_TEXT, s)
+      case FileBody(f, _)         => HttpRequest.of(method, path, OCTET_STREAM, Files.readAllBytes(f.toPath))
+      case ByteArrayBody(b, _)    => HttpRequest.of(method, path, OCTET_STREAM, b)
+      case InputStreamBody(is, _) => HttpRequest.of(method, path, OCTET_STREAM, is.readAllBytes())
+      case ByteBufferBody(b, _)   => HttpRequest.of(method, path, OCTET_STREAM, UTF_8.decode(b).toString)
```

Comment:
we're still doing a conversion to string here. It we have to convert at all, let's covert to a byte array

Reply:
In case of `ByteBufferBody` since `ByteBuffer` b can be readonly we cannot access it using `b.array()` or `b.duplicate().array()` since both of this calls end up with `ReadOnlyBufferException`. Decoding is the only solution i have found for that case.

Reply:
There's an util for that it turns out :) See `sttp.client3.internal.RichByteBuffer.safeRead`

Reply:
nice, it works :)